File: backend/app/jamming.py
"""EW (Electronic Warfare) jamming logic — jam behaviour selection and
jammed-drone movement updates."""

from __future__ import annotations

import math
import random

from app.config import KTS_TO_KMS
from app.models import DroneState, DroneType


# Per-type RF jam resistance probability (chance that RF link disruption has NO effect)
JAM_RESIST: dict[DroneType, float] = {
    DroneType.COMMERCIAL_QUAD: 0.0,   # Fully jammable — GPS + RF dependent
    DroneType.MICRO: 0.10,            # Mostly jammable — small GPS receiver
    DroneType.FIXED_WING: 0.40,       # Partial — may have basic autopilot
    DroneType.IMPROVISED: 0.50,       # Unknown RF dependency — coin flip
    DroneType.SHAHED: 1.0,            # RF-immune — autonomous INS nav, no RF link
}
# Default for any type not listed (e.g. bird, swarm)
_DEFAULT_JAM_RESIST = 0.50

# Per-type PNT vulnerability (drift magnitude in km/s applied when PNT-jammed)
# 0.0 = immune (INS navigation). Higher = GPS-dependent, larger drift.
PNT_VULNERABILITY: dict[DroneType, float] = {
    DroneType.COMMERCIAL_QUAD: 0.008,  # Heavy GPS reliance — significant drift
    DroneType.MICRO: 0.006,            # Small GPS receiver — moderate drift
    DroneType.FIXED_WING: 0.004,       # May have basic autopilot — light drift
    DroneType.IMPROVISED: 0.005,       # Unknown GPS dependency — moderate drift
    DroneType.SHAHED: 0.003,           # INS-primary but still GPS-aided — minor drift
}
# Default for unlisted types (ambient traffic, birds, etc.)
_DEFAULT_PNT_VULNERABILITY = 0.0  # Don't jam ambient/non-threat types


def pick_jam_behavior(drone_type: DroneType) -> str | None:
    """Pick a jammed behavior for the drone.

    Returns ``None`` if the jam fails based on per-type resistance.
    SHAHED always returns None (autonomous INS navigation, no RF link).
    """
    resist = JAM_RESIST.get(drone_type, _DEFAULT_JAM_RESIST)
    if random.random() < resist:
        return None  # Jam fails
    return random.choice(["loss_of_control", "rth", "forced_landing", "gps_spoof"])


def apply_pnt_jamming(drone_type: DroneType) -> tuple[bool, float]:
    """Determine PNT jamming effect for a drone type.

    Returns ``(effective, drift_magnitude)`` where *drift_magnitude* is the
    km/s of random nav error injected per tick.  SHAHED receives minor drift
    (GPS-aided INS), all other non-listed types are immune (drift = 0.0).
    """
    drift = PNT_VULNERABILITY.get(drone_type, _DEFAULT_PNT_VULNERABILITY)
    if drift <= 0.0:
        return False, 0.0
    # Add ±20% jitter to drift magnitude so each jam event feels distinct
    jitter = random.uniform(0.8, 1.2)
    return True, round(drift * jitter, 5)


def update_pnt_jammed_drone(
    drone: DroneState,
    tick_rate: float,
    elapsed: float,
) -> tuple[DroneState, list[dict]]:
    """Advance a PNT-jammed drone by one tick.

    Applies random navigation drift (heading perturbation) and decrements
    the PNT jam timer.  Does NOT neutralize the drone — PNT jamming degrades
    accuracy, it doesn't defeat the drone outright.

    Returns the updated ``DroneState`` and a (possibly empty) list of event dicts.
    """
    events: list[dict] = []

    remaining = drone.pnt_jammed_time_remaining - tick_rate
    if remaining <= 0:
        # PNT effect expired — clear state
        drone = drone.model_copy(update={
            "pnt_jammed": False,
            "pnt_drift_magnitude": 0.0,
            "pnt_jammed_time_remaining": 0.0,
        })
        events.append({
            "type": "event",
            "timestamp": round(elapsed, 1),
            "message": f"PNT: {drone.id.upper()} — NAV DEGRADATION CLEARED",
        })
        return drone, events

    # Apply drift: random heading perturbation proportional to drift magnitude
    drift_km = drone.pnt_drift_magnitude
    # Random direction for this tick's nav error
    drift_angle = random.uniform(0, 2 * math.pi)
    from app.config import KTS_TO_KMS  # avoid circular at module level
    new_x = drone.x + math.cos(drift_angle) * drift_km * tick_rate
    new_y = drone.y + math.sin(drift_angle) * drift_km * tick_rate

    trail = list(drone.trail)
    trail.append([round(new_x, 3), round(new_y, 3)])
    if len(trail) > 20:
        trail = trail[-20:]

    drone = drone.model_copy(update={
        "x": new_x,
        "y": new_y,
        "trail": trail,
        "pnt_jammed_time_remaining": remaining,
    })
    return drone, events


def update_jammed_drone(
    drone: DroneState,
    tick_rate: float,
    elapsed: float,
) -> tuple[DroneState, list[dict]]:
    """Advance a jammed drone by one tick.

    Returns the updated ``DroneState`` and a (possibly empty) list of event
    dicts to broadcast.
    """
    events: list[dict] = []

    # Decrement jam timer
    drone = drone.model_copy(update={
        "jammed_time_remaining": drone.jammed_time_remaining - tick_rate,
    })

    jb = drone.jammed_behavior

    if jb == "loss_of_control":
        speed_kms = drone.speed * KTS_TO_KMS * 0.5  # half-speed drift
        heading_rad = math.radians(drone.heading)
        new_x = drone.x + math.sin(heading_rad) * speed_kms * tick_rate
        new_y = drone.y + math.cos(heading_rad) * speed_kms * tick_rate
        new_alt = max(0, drone.altitude - 15 * tick_rate)
        trail = list(drone.trail)
        trail.append([round(new_x, 3), round(new_y, 3)])
        if len(trail) > 20:
            trail = trail[-20:]
        drone = drone.model_copy(update={
            "x": new_x, "y": new_y, "altitude": new_alt,
            "speed": drone.speed * 0.95, "trail": trail,
        })
        if new_alt <= 0 or drone.jammed_time_remaining <= 0:
            drone = drone.model_copy(update={
                "neutralized": True, "jammed_time_remaining": 0, "altitude": 0,
            })
            events.append({
                "type": "event",
                "timestamp": round(elapsed, 1),
                "message": f"TRACK: {drone.id.upper()} — CRASHED (loss of control)",
            })

    elif jb == "rth":
        away_angle = math.atan2(drone.y, drone.x)
        speed_kms = drone.speed * KTS_TO_KMS
        new_x = drone.x + math.cos(away_angle) * speed_kms * tick_rate
        new_y = drone.y + math.sin(away_angle) * speed_kms * tick_rate
        heading_deg = math.degrees(away_angle) % 360
        trail = list(drone.trail)
        trail.append([round(new_x, 3), round(new_y, 3)])
        if len(trail) > 20:
            trail = trail[-20:]
        drone = drone.model_copy(update={
            "x": new_x, "y": new_y, "heading": heading_deg, "trail": trail,
        })
        if math.sqrt(new_x ** 2 + new_y ** 2) > 10.0:
            drone = drone.model_copy(update={
                "neutralized": True, "jammed_time_remaining": 0,
            })
            events.append({
                "type": "event",
                "timestamp": round(elapsed, 1),
                "message": f"TRACK: {drone.id.upper()} — RTH (left area)",
            })

    elif jb == "forced_landing":
        new_alt = max(0, drone.altitude - 50 * tick_rate)
        drone = drone.model_copy(update={
            "altitude": new_alt,
            "speed": max(0, drone.speed - 5 * tick_rate),
        })
        if new_alt <= 0:
            drone = drone.model_copy(update={
                "neutralized": True, "jammed_time_remaining": 0,
                "altitude": 0, "speed": 0,
            })
            events.append({
                "type": "event",
                "timestamp": round(elapsed, 1),
                "message": f"TRACK: {drone.id.upper()} — FORCED LANDING (grounded)",
            })

    elif jb == "gps_spoof":
        spoof_heading = (
            drone.heading + random.uniform(90, 180) * random.choice([-1, 1])
        ) % 360
        heading_rad = math.radians(spoof_heading)
        speed_kms = drone.speed * KTS_TO_KMS
        new_x = drone.x + math.sin(heading_rad) * speed_kms * tick_rate
        new_y = drone.y + math.cos(heading_rad) * speed_kms * tick_rate
        trail = list(drone.trail)
        trail.append([round(new_x, 3), round(new_y, 3)])
        if len(trail) > 20:
            trail = trail[-20:]
        drone = drone.model_copy(update={
            "x": new_x, "y": new_y, "heading": spoof_heading, "trail": trail,
        })
        if math.sqrt(new_x ** 2 + new_y ** 2) > 10.0:
            drone = drone.model_copy(update={
                "neutralized": True, "jammed_time_remaining": 0,
            })
            events.append({
                "type": "event",
                "timestamp": round(elapsed, 1),
                "message": f"TRACK: {drone.id.upper()} — GPS SPOOFED (left area)",
            })

    return drone, events

File: src/game/jamming.ts
/**
 * OpenSentry jamming logic — direct port of backend/app/jamming.py
 */

import { markDroneNeutralized } from './state';
import type { DroneState } from './state';
import { KTS_TO_KMS } from './helpers';

// ---------------------------------------------------------------------------
// Jam resistance per drone type (0.0 = fully jammable, 1.0 = jam-immune)
// ---------------------------------------------------------------------------

export const JAM_RESIST: Record<string, number> = {
  commercial_quad: 0.15,
  micro: 0.20,
  fixed_wing: 0.40,
  improvised: 0.50,
  improvised_hardened: 0.70,
  shahed: 1.0,
};
const _DEFAULT_JAM_RESIST = 0.50;

// ---------------------------------------------------------------------------
// PNT vulnerability per drone type (drift magnitude per tick when jammed)
// ---------------------------------------------------------------------------

export const PNT_VULNERABILITY: Record<string, number> = {
  commercial_quad: 0.008,
  micro: 0.006,
  fixed_wing: 0.004,
  improvised: 0.005,
  improvised_hardened: 0.001,
  shahed: 0.0, // INS-primary — fully immune to PNT/GPS jamming
};
const _DEFAULT_PNT_VULNERABILITY = 0.0;

// ---------------------------------------------------------------------------
// Event interface returned by update functions
// ---------------------------------------------------------------------------

interface JamEvent {
  type: string;
  timestamp: number;
  message: string;
}

// ---------------------------------------------------------------------------
// Public functions
// ---------------------------------------------------------------------------

/**
 * Roll against jam resistance for the given drone type.
 * Returns null if the drone resists, otherwise a random jam behavior string.
 */
export function pickJamBehavior(droneType: string): string | null {
  const resist = JAM_RESIST[droneType] ?? _DEFAULT_JAM_RESIST;
  if (Math.random() < resist) return null;
  // Commercial quads and micros predominantly enter ATTI mode (weighted 2x)
  const behaviors = (droneType === 'commercial_quad' || droneType === 'micro')
    ? ['rth', 'atti_mode', 'atti_mode', 'gps_spoof']
    : ['loss_of_control', 'rth', 'forced_landing', 'gps_spoof'];
  return behaviors[Math.floor(Math.random() * behaviors.length)];
}

/**
 * Determine PNT jamming drift for the given drone type.
 * Returns [affected, driftMagnitude].
 */
export function applyPntJamming(droneType: string): [boolean, number] {
  const drift = PNT_VULNERABILITY[droneType] ?? _DEFAULT_PNT_VULNERABILITY;
  if (drift <= 0.0) return [false, 0.0];
  const jitter = 0.8 + Math.random() * 0.4; // uniform(0.8, 1.2)
  return [true, Math.round(drift * jitter * 100000) / 100000];
}

/**
 * Tick update for a PNT-jammed drone: apply random nav drift, decrement timer.
 */
export function updatePntJammedDrone(
  drone: DroneState,
  tickRate: number,
  elapsed: number,
): [DroneState, JamEvent[]] {
  const events: JamEvent[] = [];
  const remaining = drone.pnt_jammed_time_remaining - tickRate;

  if (remaining <= 0) {
    const updated: DroneState = {
      ...drone,
      pnt_jammed: false,
      pnt_drift_magnitude: 0.0,
      pnt_jammed_time_remaining: 0.0,
    };
    events.push({
      type: 'event',
      timestamp: Math.round(elapsed * 10) / 10,
      message: `PNT: ${(drone.display_label || drone.id).toUpperCase()} \u2014 NAV DEGRADATION CLEARED`,
    });
    return [updated, events];
  }

  const driftKm = drone.pnt_drift_magnitude;
  const driftAngle = Math.random() * 2 * Math.PI;
  const newX = drone.x + Math.cos(driftAngle) * driftKm * tickRate;
  const newY = drone.y + Math.sin(driftAngle) * driftKm * tickRate;

  const trail = [...drone.trail, [Math.round(newX * 1000) / 1000, Math.round(newY * 1000) / 1000]].slice(-20);

  const updated: DroneState = {
    ...drone,
    x: newX,
    y: newY,
    trail,
    pnt_jammed_time_remaining: remaining,
  };
  return [updated, events];
}

/**
 * Tick update for an RF-jammed drone. Applies the assigned jammed_behavior
 * (loss_of_control, rth, forced_landing, gps_spoof) until neutralized or timer expires.
 */
export function updateJammedDrone(
  drone: DroneState,
  tickRate: number,
  elapsed: number,
): [DroneState, JamEvent[]] {
  const events: JamEvent[] = [];
  let d: DroneState = {
    ...drone,
    jammed_time_remaining: drone.jammed_time_remaining - tickRate,
  };
  const jb = d.jammed_behavior;

  // Guard: if jammed=true but no behavior assigned, clear the jam state
  if (jb === null || jb === undefined) {
    const cleared: DroneState = {
      ...d,
      jammed: false,
      jammed_time_remaining: 0,
    };
    events.push({
      type: 'event',
      timestamp: Math.round(elapsed * 10) / 10,
      message: `TRACK: ${(d.display_label || d.id).toUpperCase()} — JAM STATE CLEARED (no behavior assigned)`,
    });
    return [cleared, events];
  }

  if (jb === 'loss_of_control') {
    const speedKms = d.speed * KTS_TO_KMS * 0.5;
    const headingRad = (d.heading * Math.PI) / 180;
    const newX = d.x + Math.cos(headingRad) * speedKms * tickRate;
    const newY = d.y + Math.sin(headingRad) * speedKms * tickRate;
    const newAlt = Math.max(0, d.altitude - 15 * tickRate);
    const trail = [...d.trail, [Math.round(newX * 1000) / 1000, Math.round(newY * 1000) / 1000]].slice(-20);
    d = { ...d, x: newX, y: newY, altitude: newAlt, speed: d.speed * 0.95, trail };
    if (newAlt <= 0 || d.jammed_time_remaining <= 0) {
      d = markDroneNeutralized(d, elapsed, { jammed_time_remaining: 0, altitude: 0 });
      events.push({
        type: 'event',
        timestamp: Math.round(elapsed * 10) / 10,
        message: `TRACK: ${(d.display_label || d.id).toUpperCase()} \u2014 CRASHED (loss of control)`,
      });
    }
  } else if (jb === 'rth') {
    const awayAngle = Math.atan2(d.y, d.x);
    const speedKms = d.speed * KTS_TO_KMS;
    const newX = d.x + Math.cos(awayAngle) * speedKms * tickRate;
    const newY = d.y + Math.sin(awayAngle) * speedKms * tickRate;
    const headingDeg = ((awayAngle * 180) / Math.PI + 360) % 360;
    const trail = [...d.trail, [Math.round(newX * 1000) / 1000, Math.round(newY * 1000) / 1000]].slice(-20);
    d = { ...d, x: newX, y: newY, heading: headingDeg, trail };
    if (Math.sqrt(newX ** 2 + newY ** 2) > 10.0 || d.jammed_time_remaining <= 0) {
      d = markDroneNeutralized(d, elapsed, { jammed_time_remaining: 0 });
      events.push({
        type: 'event',
        timestamp: Math.round(elapsed * 10) / 10,
        message: `TRACK: ${(d.display_label || d.id).toUpperCase()} \u2014 RTH (left area)`,
      });
    }
  } else if (jb === 'forced_landing') {
    const newAlt = Math.max(0, d.altitude - 50 * tickRate);
    d = { ...d, altitude: newAlt, speed: Math.max(0, d.speed - 5 * tickRate) };
    if (newAlt <= 0) {
      d = markDroneNeutralized(d, elapsed, { jammed_time_remaining: 0, altitude: 0, speed: 0 });
      events.push({
        type: 'event',
        timestamp: Math.round(elapsed * 10) / 10,
        message: `TRACK: ${(d.display_label || d.id).toUpperCase()} \u2014 FORCED LANDING (grounded)`,
      });
    }
  } else if (jb === 'atti_mode') {
    // ATTI mode: drone maintains heading/speed with small random lateral drift
    const speedKms = d.speed * KTS_TO_KMS;
    const headingRad = (d.heading * Math.PI) / 180;
    const newX = d.x + Math.cos(headingRad) * speedKms * tickRate + (Math.random() - 0.5) * 0.004;
    const newY = d.y + Math.sin(headingRad) * speedKms * tickRate + (Math.random() - 0.5) * 0.004;
    const trail = [...d.trail, [Math.round(newX * 1000) / 1000, Math.round(newY * 1000) / 1000]].slice(-20);
    d = { ...d, x: newX, y: newY, trail };
    // ATTI mode does NOT neutralize — clears when timer expires
    // jam_cooldown prevents tickPassiveJamming from immediately re-jamming
    if (d.jammed_time_remaining <= 0) {
      d = { ...d, jammed: false, jammed_behavior: null, jammed_time_remaining: 0, jam_cooldown: 15.0 };
      events.push({
        type: 'event',
        timestamp: Math.round(elapsed * 10) / 10,
        message: `EW: ${(d.display_label || d.id).toUpperCase()} \u2014 ATTI MODE CLEARED, RESUMING AUTONOMOUS NAVIGATION`,
      });
    }
  } else if (jb === 'gps_spoof') {
    const sign = Math.random() < 0.5 ? -1 : 1;
    const spoofHeading = (d.heading + (90 + Math.random() * 90) * sign + 360) % 360;
    const headingRad = (spoofHeading * Math.PI) / 180;
    const speedKms = d.speed * KTS_TO_KMS;
    const newX = d.x + Math.cos(headingRad) * speedKms * tickRate;
    const newY = d.y + Math.sin(headingRad) * speedKms * tickRate;
    const trail = [...d.trail, [Math.round(newX * 1000) / 1000, Math.round(newY * 1000) / 1000]].slice(-20);
    d = { ...d, x: newX, y: newY, heading: spoofHeading, trail };
    if (Math.sqrt(newX ** 2 + newY ** 2) > 10.0 || d.jammed_time_remaining <= 0) {
      d = markDroneNeutralized(d, elapsed, { jammed_time_remaining: 0 });
      events.push({
        type: 'event',
        timestamp: Math.round(elapsed * 10) / 10,
        message: `TRACK: ${(d.display_label || d.id).toUpperCase()} \u2014 GPS SPOOFED (left area)`,
      });
    }
  }

  return [d, events];
}
